### PR TITLE
953 fix patient recreate at HIEs

### PR DESCRIPTION
**Pull request by leite08**
Ref: metriport/metriport#953

### Dependencies

- Upstream: https://github.com/metriport/metriport/pull/780
- Downstream: none

### Description

- findOrCreate @ CW to reenrol person if unenrolled
- return earliest enrolled person instead of latest
- keep recreating at HIEs when 1 patient fails

### Release Plan

- asap

### CODE REVIEW

File: packages/api/src/external/commonwell/patient-shared.ts [hunk 1]
```diff
@@ -70,27 +72,56 @@ export async function findOrCreatePerson({
...
-      ? respSearch._embedded.person.flatMap(p => (p && getId(p) ? p : []))
+      ? respSearch._embedded.person
+          .flatMap(p => (p && getPersonId(p) ? p : []))
+          .flatMap(filterTruthy)
       : [];
-    if (persons.length > 1) {
-      return alertAndReturnMostRecentPerson(
+
+    const enrolledPersons = persons.filter(isEnrolled);
+    if (enrolledPersons.length === 1) {
```
Comment (Goncharo) [inline packages/api/src/external/commonwell/patient-shared.ts:81]: What about the === 0 case?

File: packages/api/src/external/commonwell/patient-shared.ts [hunk 2]
```diff
@@ -70,27 +72,56 @@ export async function findOrCreatePerson({
...
+      return alertAndReturnEarliestPerson(
+        commonwellPatientId,
+        [enrolledPersons[0] as CommonwellPerson, ...enrolledPersons.slice(1)], // to match the type requiring at least one element
+        commonWell.lastReferenceHeader,
+        context
+      );
+    }
+
+    const unenrolledPersons = persons.filter(isUnenrolled);
+    if (unenrolledPersons.length === 1) {
```
Comment (Goncharo) [inline packages/api/src/external/commonwell/patient-shared.ts:97]: ditto

**Review by Goncharo — APPROVED** [review]
🚢 - some minor nits/comments

File: packages/api/src/external/commonwell/patient-shared.ts (hunk 1 above)
Comment (leite08) [reply packages/api/src/external/commonwell/patient-shared.ts:81]: then it tries unenrolled

File: packages/api/src/external/commonwell/patient-shared.ts (hunk 2 above)
Comment (leite08) [reply packages/api/src/external/commonwell/patient-shared.ts:97]: then it adds/enrolls

File: packages/api/src/external/commonwell/admin/recreate-patients-at-hies.ts [hunk 3]
```diff
@@ -60,81 +60,88 @@ export async function recreatePatientAtCW(
...
-  const originalCWPatientId = commonwellData.patientId;
-  if (originalCWPatientId.includes(patient.id)) {
-    const msg = `Patient ID and CW patientId already match`;
-    log(msg + ", skipping...");
-    capture.message(msg, { extra: { patient }, level: "warning" });
-    return undefined;
-  }
+  try {
+    const facilityId = patient.facilityIds[0];
+    if (!facilityId) {
```
[the remark this answers is not in the document]
Comment (leite08) [reply packages/api/src/external/commonwell/admin/recreate-patients-at-hies.ts:65]: Agree it shouldn't happen, but I'd not interrupt processing.

Note: this was already like this, its showing as a change but I just wrapped this function's code within a try/catch.

**Pull request merged**